components/script/animations.rs

```diff
@@ -171,103 +204,195 @@ impl Animations {
+                TransitionOrAnimationEventType::TransitionRun => atom!("transitionrun"),
+            };
+            let parent = EventInit {
+                bubbles: true,
+                cancelable: false,
+            };
+
+            // TODO: Handle pseudo-elements properly
+            let property_or_animation_name =
+                DOMString::from(event.property_or_animation_name.clone());
+            let elapsed_time = Finite::new(event.elapsed_time as f32).unwrap();
+            let window = window_from_node(&*node);
+
+            if event.event_type.is_transition_event() {
+                let event_init = TransitionEventInit {
+                    parent,
+                    propertyName: property_or_animation_name,
+                    elapsedTime: elapsed_time,
+                    pseudoElement: DOMString::new(),
+                };
+                TransitionEvent::new(&window, event_atom, &event_init)
+                    .upcast::<Event>()
+                    .fire(node.upcast());
+            } else {
+                let event_init = AnimationEventInit {
+                    parent,
+                    animationName: property_or_animation_name,
+                    elapsedTime: elapsed_time,
+                    pseudoElement: DOMString::new(),
+                };
+                AnimationEvent::new(&window, event_atom, &event_init)
+                    .upcast::<Event>()
+                    .fire(node.upcast());
+            }
+        }
     }
+}
 
-    /// Collect newly animating nodes, which is used by the script process during
-    /// forced, synchronous reflows to root DOM nodes for the duration of their
-    /// animations or transitions.
-    /// TODO(mrobinson): Look into handling the rooting inside this class.
-    fn collect_newly_animating_nodes(
-        &mut self,
-        animation_states: &FxHashMap<OpaqueNode, ElementAnimationSet>,
-    ) {
-        // This extends the output vector with an iterator that contains a copy of the node
-        // address for every new animation. The script thread currently stores a rooted node
-        // for every property that is transitioning. The current strategy of repeating the
-        // node address is a holdover from when the code here looked different.
-        self.newly_animating_nodes
-            .extend(animation_states.iter().flat_map(|(node, state)| {
-                let mut num_new_animations = state
-                    .animations
-                    .iter()
-                    .filter(|animation| animation.is_new)
-                    .count();
-                num_new_animations += state
-                    .transitions
-                    .iter()
-                    .filter(|transition| transition.is_new)
-                    .count();
-
-                let node = UntrustedNodeAddress(node.0 as *const c_void);
-                std::iter::repeat(node).take(num_new_animations)
-            }));
+impl MallocSizeOf for Animations {
```

Comment:
Can we derive this instead and use the `#[ignore_malloc_size_of]` annotation where necessary?

Reply:
Sure. This means we won't measure the size of the animation sets, but perhaps it makes sense to add a new directive to `malloc_size_of_derive` to make it easier to fix this in a foolproof way.